examples/sdxl.py

```diff
@@ -0,0 +1,1107 @@
+      x = x + self.mlp(self.ln_2(x))
+      return x
+
+
+  # https://github.com/mlfoundations/open_clip/blob/58e4e39aaabc6040839b0d2a7e8bf20979e4558a/src/open_clip/transformer.py#L317
+  class ClipTransformer:
+    def __init__(self, dims:int, layers:int, n_heads:int, mlp_ratio:float=4.0):
+      self.resblocks = [
+        Open.ResidualAttentionBlocks(dims, n_heads, mlp_ratio) for _ in range(layers)
+      ]
+
+    def __call__(self, x:Tensor, attn_mask:Optional[Tensor]=None) -> Tensor:
+      x = x.transpose(0, 1).contiguous()
+      for r in self.resblocks:
+        x = r(x, attn_mask=attn_mask)
+      x = x.transpose(0, 1)
+      return x
+
+
+  # https://github.com/mlfoundations/open_clip/blob/58e4e39aaabc6040839b0d2a7e8bf20979e4558a/src/open_clip/model.py#L220
+  # https://github.com/mlfoundations/open_clip/blob/58e4e39aaabc6040839b0d2a7e8bf20979e4558a/src/open_clip/transformer.py#L661
+  class ClipTextTransformer:
+    def __init__(self, dims:int, vocab_size:int=49408, n_heads:int=20, ctx_length:int=77, layers:int=32):
+      self.token_embedding = Embedding(vocab_size, dims)
+      self.positional_embedding = Tensor.empty(ctx_length, dims)
+      self.transformer = Open.ClipTransformer(dims, layers, n_heads)
+      self.ln_final = LayerNorm(dims)
+      self.text_projection = Tensor.empty(dims, dims)
+
+    @property
+    def attn_mask(self) -> Tensor:
+      if not hasattr(self, "_attn_mask"):
+        self._attn_mask = Tensor.full((77, 77), float("-inf")).triu(1)
+      return self._attn_mask
+
+    def __call__(self, text:Tensor) -> Tensor:
+      seq_len = text.shape[1]
+
+      x = self.token_embedding(text)
+      x = x + self.positional_embedding[:seq_len]
+      x = self.transformer(x, attn_mask=self.attn_mask)
+      x = self.ln_final(x)
+
+      pooled = x[Tensor.arange(x.shape[0]), text.argmax(dim=-1)]
+      pooled = pooled @ self.text_projection
+      return pooled
+
+
+# https://github.com/Stability-AI/generative-models/blob/fbdc58cab9f4ee2be7a5e1f2e2787ecd9311942f/sgm/modules/encoders/modules.py#L396
+class FrozenOpenClipEmbedder(Embedder):
+  def __init__(self, dims:int=1280):
+    self.model = Open.ClipTextTransformer(dims)
+    self.input_key = "txt"
+    self.tokenizer = Tokenizer.ClipTokenizer()
+
+  def text_transformer_forward(self, x:Tensor, attn_mask:Optional[Tensor]=None):
+    for r in self.model.transformer.resblocks:
+      x, penultimate = r(x, attn_mask=attn_mask), x
+    return x.permute(1,0,2), penultimate.permute(1,0,2)
+
+  def __call__(self, text:Tensor) -> Tensor:
+    tokens = Tensor(self.tokenizer.encode(text, pad_with_zeros=True), dtype=dtypes.int64).reshape(1,-1)
+
+    x = self.model.token_embedding(tokens).add(self.model.positional_embedding).permute(1,0,2)
+    x, penultimate = self.text_transformer_forward(x, attn_mask=self.model.attn_mask)
+    x = self.model.ln_final(x)
+    pooled = x[Tensor.arange(x.shape[0]), tokens.argmax(axis=-1).numpy().item()] @ self.model.text_projection
```

Comment:
is this `numpy()` needed?

Reply:
Gave me problems originally, but apparently works without now, removed in 4aef32d